[Bugfix]: v1 engine - consider lora adapters in allowed_token_ids

bbrowning opened the pull request:
Previously, the v1 engine was only look at the base model's tokenizer to determine the range of token id values that were valid in allowed_token_ids. Now, it takes into account any additional tokens added by any lora adapters in use for this request.

The existing v0 behavior already took this into account, so this just closes that gap for the v1 engine.

I added a unit test to exercise this as well, where the test creates the minimal fake lora adapter files on disk using the pytest `tmp_path` fixture and then ensures the tokenizer we created for that fake lora adapter is getting used when determining the allowed_token_ids valid values. The test failed before this change and passes afterwards.

FIX #17746

bbrowning committed:
[Bugfix]: v1 engine - consider lora adapters in allowed_token_ids

Previously, the v1 engine was only look at the base model's tokenizer
to determine the range of token id values that were valid in
allowed_token_ids. Now, it takes into account any additional tokens
added by any lora adapters in use for this request.

The existing v0 behavior already took this into account, so this just
closes that gap for the v1 engine.

I added a unit test to exercise this as well, where the test
creates the minimal fake lora adapter files on disk using the pytest
`tmp_path` fixture and then ensures the tokenizer we created for that
fake lora adapter is getting used when determining the
allowed_token_ids valid values. The test failed before this change and
passes afterwards.

Fixes #17746

Signed-off-by: Ben Browning <bbrownin@redhat.com>

jeejeelee commented:
Thank you for your contribution, I will look at this ASAP

jeejeelee commented on tests/v1/engine/test_processor.py:
```diff
@@ -0,0 +1,193 @@
+
+@pytest.fixture
+def fake_lora_adapter_files(tmp_path, opt_125m_huggingface_id):
+    adapter_config = {
```
We can use this [real adapter](https://github.com/vllm-project/vllm/blob/main/tests/lora/conftest.py#L143) to complete the test.
Maybe we should rename `test_processor.py` to `test_lora_allow_token_ids.py` and move it to the `tests/lora/`directory

bbrowning replied:
I swapped to the real adapter, which required me to find a real ungated base model for that adapter since we have to also have the base model's tokenizer around. I chose `RedHatAI/Llama2-7b-chat-pruned50-quant-ds` as an ungated base model for the sql lora adapter, as the quantization doesn't really matter so much as just having a value tokenizer config in that base model.

jeejeelee commented on vllm/v1/engine/processor.py:
```diff
@@ -82,7 +83,8 @@ def _validate_sampling_params(
         if not params.allowed_token_ids:
             raise ValueError("allowed_token_ids is not None and empty!")
-        vocab_size = self.model_config.get_vocab_size()
+        tokenizer = self.tokenizer.get_lora_tokenizer(lora_request)
```
Many LoRA adapters don't contain tokenizer, so this will return None, we need to consider this

bbrowning replied:
The logic behind `self.tokenizer.get_lora_tokenizer(lora_request)` already gracefully handles this case, returning the base model's tokenizer if the adapter does not have its own. However, I will expand the testing here (along with moving it as you suggested above) to cover that case to ensure it doesn't regress.

bbrowning replied:
I added a second test method to this suite that uses the `jeeejeee/qwen25-vl-lora-pokemon` adapter, as that's a good example of one that does not define additional vocabulary or bring its own tokenizer. This second test ensures adapters that do not define their own tokenizer continue to work.

bbrowning committed:
chore: expand lora allowed_token_ids testing

This expands the scope of the `allowed_token_ids` testing for LoRA
adapters, testing both for adapters that define additional tokens and
adapters that do not. It also tests that we still properly raise
errors for token values passed in that do not fall within the expected
base and/or adapter's valid token ranges.

Lastly, it moves the test from tests/v1/engine/test_processor.py to
tests/lora/test_lora_allowed_token_ids.py to take advantage of the
lora adapter fixtures already defined in tests/lora/conftest.py as
well as adding a couple new ones there to cover these testing needs.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

jeejeelee commented on tests/lora/conftest.py:
```diff
@@ -139,6 +139,12 @@ def dummy_model_gate_up() -> nn.Module:
+@pytest.fixture(scope="session")
+def llama_2_7b_base_huggingface_id():
+    # used as an ungated base model for testing with sql lora adapter
+    return "RedHatAI/Llama2-7b-chat-pruned50-quant-ds"
```
QQ: Will this download the model? If so, I suggest changing it to `meta-llama/Llama-2-7b-hf`

bbrowning replied:
As used in this test, it downloads the tokenizer, but not the model weights. I'm happy to change it to the meta-llama variant, but that one is gated. Are gated models allowed in the tests, and CI is run with valid huggingface tokens with access to gated Llama 2 models? All the other lora examples look like ungated ones. The only reason I picked this one was to get something with an identical tokenizer to llama2-7b but ungated.

Here's what it downloaded locally when I ran these unit tests:

```
ls -1 ~/.cache/huggingface/hub/models--RedHatAI--Llama2-7b-chat-pruned50-quant-ds/snapshots/e54ea3738c076a09cd6124971534b956ab40b8cd 
config.json
special_tokens_map.json
tokenizer_config.json
tokenizer.json
tokenizer.model
```

jeejeelee replied:
CI can access to `meta-llama/Llama-2-7b-hf`, see: https://github.com/vllm-project/vllm/blob/v0.8.5/tests/lora/test_llama_tp.py#L11 .
 It's still better to change it to `meta-llama/Llama-2-7b-hf` to avoid potential future tests that might download this model.

bbrowning replied:
Great - updated to the standard meta-llama base model in that case. Thanks!

jeejeelee reviewed (approved):
LGTM, thank you
